Sources/Nats/BatchBuffer.swift

```diff
@@ -15,124 +15,113 @@ import Foundation
 import NIO
 import NIOConcurrencyHelpers
 
-class BatchBuffer {
+extension BatchBuffer {
+    struct State {
+        private var buffer: ByteBuffer
+        private var allocator: ByteBufferAllocator
+        var waitingPromises: [(ClientOp, UnsafeContinuation<Void, Error>)] = []
+        var isWriteInProgress: Bool = false
+
+        internal init(allocator: ByteBufferAllocator, batchSize: Int = 16 * 1024) {
+            self.allocator = allocator
+            self.buffer = allocator.buffer(capacity: batchSize)
+        }
+
+        var readableBytes: Int {
+            return self.buffer.readableBytes
+        }
+
+        mutating func getWriteBuffer() -> ByteBuffer {
+            var writeBuffer = allocator.buffer(capacity: buffer.readableBytes)
+            writeBuffer.writeBytes(buffer.readableBytesView)
+            buffer.clear()
+
+            return writeBuffer
+        }
+
+        mutating func writeMessage(_ message: ClientOp) {
+            self.buffer.writeClientOp(message)
+        }
+    }
+}
+
+internal class BatchBuffer {
     private let batchSize: Int
-    private var buffer: ByteBuffer
     private let channel: Channel
-    private let lock = NIOLock()
-    private var waitingPromises: [EventLoopPromise<Void>] = []
-    private var isWriteInProgress: Bool = false
+    private let state: NIOLockedValueBox<State>
 
     init(channel: Channel, batchSize: Int = 16 * 1024) {
         self.batchSize = batchSize
-        self.buffer = channel.allocator.buffer(capacity: batchSize)
         self.channel = channel
+        self.state = .init(
+            State(allocator: channel.allocator)
+        )
     }
 
-    func write<Bytes: Sequence>(_ data: Bytes) async throws where Bytes.Element == UInt8 {
+    func writeMessage(_ message: ClientOp) async throws {
```

Comment:
is there a reason why we have this func with exactly the same signature as write()? can we rename the write just below?